core/translate.c

```diff
@@ -189,11 +189,35 @@ translate_walk_track(dcontext_t *tdcontext, instr_t *inst, translate_walk_t *wal
         walk->unsupported_mangle = false;
         walk->xsp_adjust = 0;
         for (r = 0; r < REG_SPILL_NUM; r++) {
+#ifndef ARM
             /* we should have seen a restore for every spill, unless at
              * fragment-ending jump to ibl, which shouldn't come here
              */
             ASSERT(!walk->reg_spilled[r]);
             walk->reg_spilled[r] = false; /* be paranoid */
+#else
+            /* On ARM we do spill registers across app instrs and mangle
+             * regions, though right now only the following routines do this:
+             * - mangle_stolen_reg()
+             * - mangle_gpr_list_read()
+             * - mangle_reads_thread_register()
+             * Each of these cases is a tls restore, and we assert as much.
+             */
+            DOCHECK(1, {
```

Comment:
> Should I make this a separate CL?

That would be cleaner but of course there's overhead to doing that, and given that nobody hit this before and it's not already in the tracker I won't complain if you keep it here.

> Should we lift all this out of the DOCHECK?
> Is an instrlist walk too expensive, even for -debug?

No, -checklevel 1 is the default, so it will run in a default debug run, and we want -checklevel 0 to remove some checks.

> Are these three mangle functions really the only ones that reserve over an app instruction? I didn't audit all the code thoroughly, just checked for comments noting "app instruction executes here..."

Probably, and I think the API restrictions on clients doing stuff across app instrs still apply, so in some sense these mangling routines are special-cased.